README.rst
rotate-backups: Simple command line interface for backup rotation
=================================================================

.. image:: https://travis-ci.org/xolox/python-rotate-backups.svg?branch=master
   :target: https://travis-ci.org/xolox/python-rotate-backups

.. image:: https://coveralls.io/repos/xolox/python-rotate-backups/badge.svg?branch=master
   :target: https://coveralls.io/r/xolox/python-rotate-backups?branch=master

Backups are good for you. Most people learn this the hard way (including me).
Nowadays my Linux laptop automatically creates a full system snapshot every
four hours by pushing changed files to an `rsync`_ daemon running on the server
in my home network and creating a snapshot afterwards using the ``cp -al``
command (the article `Easy Automated Snapshot-Style Backups with Linux and
Rsync`_ explains the basic technique). The server has a second disk attached
which asynchronously copies from the main disk so that a single disk failure
doesn't wipe all of my backups (the "time delayed replication" aspect has also
proven to be very useful).

Okay, cool, now I have backups of everything, up to date and going back in
time! But I'm running through disk space like crazy... A proper deduplicating
filesystem would be awesome but I'm running crappy consumer grade hardware and
e.g. ZFS has not been a good experience in the past. So I'm going to have to
delete backups...

Deleting backups is never nice, but an easy and proper rotation scheme can help
a lot. I wanted to keep things manageable so I wrote a Python script to do it
for me. Over the years I actually wrote several variants. Because I kept
copy/pasting these scripts around I decided to bring the main features together
in a properly documented Python package and upload it to the `Python Package
Index`_.

The `rotate-backups` package is currently tested on cPython 2.7, 3.5+ and PyPy
(2.7). It's tested on Linux and Mac OS X and may work on other unixes but
definitely won't work on Windows right now.

.. contents::
   :local:

Features
--------

**Dry run mode**
  **Use it.** I'm serious. If you don't and `rotate-backups` eats more backups
  than intended you have no right to complain ;-)

**Flexible rotation**
  Rotation with any combination of hourly, daily, weekly, monthly and yearly
  retention periods.

**Fuzzy timestamp matching in filenames**
  The modification times of the files and/or directories are not relevant. If
  you speak Python regular expressions, here is how the fuzzy matching
  works::

   # Required components.
   (?P<year>\d{4}) \D?
   (?P<month>\d{2}) \D?
   (?P<day>\d{2}) \D?
   (
      # Optional components.
      (?P<hour>\d{2}) \D?
      (?P<minute>\d{2}) \D?
      (?P<second>\d{2})?
   )?

**All actions are logged**
  Log messages are saved to the system log (e.g. ``/var/log/syslog``) so you
  can retrace what happened when something seems to have gone wrong.

Installation
------------

The `rotate-backups` package is available on PyPI_ which means installation
should be as simple as:

.. code-block:: sh

   $ pip install rotate-backups

There's actually a multitude of ways to install Python packages (e.g. the `per
user site-packages directory`_, `virtual environments`_ or just installing
system wide) and I have no intention of getting into that discussion here, so
if this intimidates you then read up on your options before returning to these
instructions ;-).

Usage
-----

There are two ways to use the `rotate-backups` package: As the command line
program ``rotate-backups`` and as a Python API. For details about the Python
API please refer to the API documentation available on `Read the Docs`_. The
command line interface is described below.

Command line
~~~~~~~~~~~~

.. A DRY solution to avoid duplication of the `rotate-backups --help' text:
..
.. [[[cog
.. from humanfriendly.usage import inject_usage
.. inject_usage('rotate_backups.cli')
.. ]]]

**Usage:** `rotate-backups [OPTIONS] [DIRECTORY, ..]`

Easy rotation of backups based on the Python package by the same name.

To use this program you specify a rotation scheme via (a combination of) the
``--hourly``, ``--daily``, ``--weekly``, ``--monthly`` and/or ``--yearly`` options and the
directory (or directories) containing backups to rotate as one or more
positional arguments.

You can rotate backups on a remote system over SSH by prefixing a DIRECTORY
with an SSH alias and separating the two with a colon (similar to how rsync
accepts remote locations).

Instead of specifying directories and a rotation scheme on the command line you
can also add them to a configuration file. For more details refer to the online
documentation (see also the ``--config`` option).

Please use the ``--dry-run`` option to test the effect of the specified rotation
scheme before letting this program loose on your precious backups! If you don't
test the results using the dry run mode and this program eats more backups than
intended you have no right to complain ;-).

**Supported options:**

.. csv-table::
   :header: Option, Description
   :widths: 30, 70


   "``-M``, ``--minutely=COUNT``","In a literal sense this option sets the number of ""backups per minute"" to
   preserve during rotation. For most use cases that doesn't make a lot of
   sense :-) but you can combine the ``--minutely`` and ``--relaxed`` options to
   preserve more than one backup per hour.  Refer to the usage of the ``-H``,
   ``--hourly`` option for details about ``COUNT``."
   "``-H``, ``--hourly=COUNT``","Set the number of hourly backups to preserve during rotation:
   
   - If ``COUNT`` is a number it gives the number of hourly backups to preserve,
     starting from the most recent hourly backup and counting back in time.
   - Alternatively you can provide an expression that will be evaluated to get
     a number (e.g. if ``COUNT`` is ""7 \* 2"" the result would be 14).
   - You can also pass ""always"" for ``COUNT``, in this case all hourly backups are
     preserved.
   - By default no hourly backups are preserved."
   "``-d``, ``--daily=COUNT``","Set the number of daily backups to preserve during rotation. Refer to the
   usage of the ``-H``, ``--hourly`` option for details about ``COUNT``."
   "``-w``, ``--weekly=COUNT``","Set the number of weekly backups to preserve during rotation. Refer to the
   usage of the ``-H``, ``--hourly`` option for details about ``COUNT``."
   "``-m``, ``--monthly=COUNT``","Set the number of monthly backups to preserve during rotation. Refer to the
   usage of the ``-H``, ``--hourly`` option for details about ``COUNT``."
   "``-y``, ``--yearly=COUNT``","Set the number of yearly backups to preserve during rotation. Refer to the
   usage of the ``-H``, ``--hourly`` option for details about ``COUNT``."
   "``-t``, ``--timestamp-pattern=PATTERN``","Customize the regular expression pattern that is used to match and extract
   timestamps from filenames. ``PATTERN`` is expected to be a Python compatible
   regular expression that must define the named capture groups 'year',
   'month' and 'day' and may define 'hour', 'minute' and 'second'."
   "``-I``, ``--include=PATTERN``","Only process backups that match the shell pattern given by ``PATTERN``. This
   argument can be repeated. Make sure to quote ``PATTERN`` so the shell doesn't
   expand the pattern before it's received by rotate-backups."
   "``-x``, ``--exclude=PATTERN``","Don't process backups that match the shell pattern given by ``PATTERN``. This
   argument can be repeated. Make sure to quote ``PATTERN`` so the shell doesn't
   expand the pattern before it's received by rotate-backups."
   "``-j``, ``--parallel``","Remove backups in parallel, one backup per mount point at a time. The idea
   behind this approach is that parallel rotation is most useful when the
   files to be removed are on different disks and so multiple devices can be
   utilized at the same time.
   
   Because mount points are per system the ``-j``, ``--parallel`` option will also
   parallelize over backups located on multiple remote systems."
   "``-p``, ``--prefer-recent``","By default the first (oldest) backup in each time slot is preserved. If
   you'd prefer to keep the most recent backup in each time slot instead then
   this option is for you."
   "``-r``, ``--relaxed``","By default the time window for each rotation scheme is enforced (this is
   referred to as strict rotation) but the ``-r``, ``--relaxed`` option can be used
   to alter this behavior. The easiest way to explain the difference between
   strict and relaxed rotation is using an example:
   
   - When using strict rotation and the number of hourly backups to preserve
     is three, only backups created in the relevant time window (the hour of
     the most recent backup and the two hours leading up to that) will match
     the hourly frequency.
   
   - When using relaxed rotation the three most recent backups will all match
     the hourly frequency (and thus be preserved), regardless of the
     calculated time window.
   
   If the explanation above is not clear enough, here's a simple way to decide
   whether you want to customize this behavior or not:
   
   - If your backups are created at regular intervals and you never miss an
     interval then strict rotation (the default) is probably the best choice.
   
   - If your backups are created at irregular intervals then you may want to
     use the ``-r``, ``--relaxed`` option in order to preserve more backups."
   "``-i``, ``--ionice=CLASS``","Use the ""ionice"" program to set the I/O scheduling class and priority of
   the ""rm"" invocations used to remove backups. ``CLASS`` is expected to be one of
   the values ""idle"" (3), ""best-effort"" (2) or ""realtime"" (1). Refer to the
   man page of the ""ionice"" program for details about these values. The
   numeric values are required by the 'busybox' implementation of 'ionice'."
   "``-c``, ``--config=FILENAME``","Load configuration from ``FILENAME``. If this option isn't given the following
   default locations are searched for configuration files:
   
   - /etc/rotate-backups.ini and /etc/rotate-backups.d/\*.ini
   - ~/.rotate-backups.ini and ~/.rotate-backups.d/\*.ini
   - ~/.config/rotate-backups.ini and ~/.config/rotate-backups.d/\*.ini
   
   Any available configuration files are loaded in the order given above, so
   that sections in user-specific configuration files override sections by the
   same name in system-wide configuration files. For more details refer to the
   online documentation."
   "``-C``, ``--removal-command=CMD``","Change the command used to remove backups. The value of ``CMD`` defaults to
   ``rm ``-f``R``. This choice was made because it works regardless of whether
   ""backups to be rotated"" are files or directories or a mixture of both.
   
   As an example of why you might want to change this, CephFS snapshots are
   represented as regular directory trees that can be deleted at once with a
   single 'rmdir' command (even though according to POSIX semantics this
   command should refuse to remove nonempty directories, but I digress)."
   "``-u``, ``--use-sudo``","Enable the use of ""sudo"" to rotate backups in directories that are not
   readable and/or writable for the current user (or the user logged in to a
   remote system over SSH)."
   "``-S``, ``--syslog=CHOICE``","Explicitly enable or disable system logging instead of letting the program
   figure out what to do. The values '1', 'yes', 'true' and 'on' enable system
   logging whereas the values '0', 'no', 'false' and 'off' disable it."
   "``-f``, ``--force``","If a sanity check fails an error is reported and the program aborts. You
   can use ``--force`` to continue with backup rotation instead. Sanity checks
   are done to ensure that the given DIRECTORY exists, is readable and is
   writable. If the ``--removal-command`` option is given then the last sanity
   check (that the given location is writable) is skipped (because custom
   removal commands imply custom semantics)."
   "``-n``, ``--dry-run``","Don't make any changes, just print what would be done. This makes it easy
   to evaluate the impact of a rotation scheme without losing any backups."
   "``-v``, ``--verbose``",Increase logging verbosity (can be repeated).
   "``-q``, ``--quiet``",Decrease logging verbosity (can be repeated).
   "``-h``, ``--help``",Show this message and exit.

.. [[[end]]]

Configuration files
~~~~~~~~~~~~~~~~~~~

Instead of specifying directories and rotation schemes on the command line you
can also add them to a configuration file.

.. [[[cog
.. from update_dotdee import inject_documentation
.. inject_documentation(program_name='rotate-backups')
.. ]]]

Configuration files are text files in the subset of `ini syntax`_ supported by
Python's configparser_ module. They can be located in the following places:

=========  ============================  =================================
Directory  Main configuration file       Modular configuration files
=========  ============================  =================================
/etc       /etc/rotate-backups.ini       /etc/rotate-backups.d/\*.ini
~          ~/.rotate-backups.ini         ~/.rotate-backups.d/\*.ini
~/.config  ~/.config/rotate-backups.ini  ~/.config/rotate-backups.d/\*.ini
=========  ============================  =================================

The available configuration files are loaded in the order given above, so that
user specific configuration files override system wide configuration files.

.. _configparser: https://docs.python.org/3/library/configparser.html
.. _ini syntax: https://en.wikipedia.org/wiki/INI_file

.. [[[end]]]

You can load a configuration file in a nonstandard location using the command
line option ``--config``, in this case the default locations mentioned above
are ignored.

Each section in the configuration defines a directory that contains backups to
be rotated. The options in each section define the rotation scheme and other
options. Here's an example based on how I use `rotate-backups` to rotate the
backups of the Linux installations that I make regular backups of:

.. code-block:: ini

   # /etc/rotate-backups.ini:
   # Configuration file for the rotate-backups program that specifies
   # directories containing backups to be rotated according to specific
   # rotation schemes.

   [/backups/laptop]
   hourly = 24
   daily = 7
   weekly = 4
   monthly = 12
   yearly = always
   ionice = idle

   [/backups/server]
   daily = 7 * 2
   weekly = 4 * 2
   monthly = 12 * 4
   yearly = always
   ionice = idle

   [/backups/mopidy]
   daily = 7
   weekly = 4
   monthly = 2
   ionice = idle

   [/backups/xbmc]
   daily = 7
   weekly = 4
   monthly = 2
   ionice = idle

As you can see in the retention periods of the directory ``/backups/server`` in
the example above you are allowed to use expressions that evaluate to a number
(instead of having to write out the literal number).

Here's an example of a configuration for two remote directories:

.. code-block:: ini

   # SSH as a regular user and use `sudo' to elevate privileges.
   [server:/backups/laptop]
   use-sudo = yes
   hourly = 24
   daily = 7
   weekly = 4
   monthly = 12
   yearly = always
   ionice = idle

   # SSH as the root user (avoids sudo passwords).
   [server:/backups/server]
   ssh-user = root
   hourly = 24
   daily = 7
   weekly = 4
   monthly = 12
   yearly = always
   ionice = idle

As this example shows you have the option to connect as the root user or to
connect as a regular user and use ``sudo`` to elevate privileges.

Customizing the rotation algorithm
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Since publishing `rotate-backups` I've found that the default rotation
algorithm is not to everyone's satisfaction and because the suggested
alternatives were just as valid as the choices that I initially made,

CHANGELOG.rst
Changelog
=========

The purpose of this document is to list all of the notable changes to this
project. The format was inspired by `Keep a Changelog`_. This project adheres
to `semantic versioning`_.

.. contents::
   :local:

.. _Keep a Changelog: http://keepachangelog.com/
.. _semantic versioning: http://semver.org/

`Release 8.1`_ (2020-05-17)
---------------------------

- Bug fix to really make the 'hour', 'minute' and 'second' capture groups in
  user defined timestamp patterns optional (this fixes issue `#26`_).

- Fixed :man:`humanfriendly` 8 deprecation warnings.

.. _Release 8.1: https://github.com/xolox/python-rotate-backups/compare/8.0...8.1
.. _#26: https://github.com/xolox/python-rotate-backups/issues/26

`Release 8.0`_ (2020-02-18)
---------------------------

This is a bit of an awkward release:

- An :exc:`~exceptions.ImportError` was reported in issue `#24`_ caused by a
  backwards incompatible change in :pypi:`humanfriendly` concerning an
  undocumented module level variable (shouldn't have used that).

- I've now updated :pypi:`rotate-backups` to be compatible with the newest
  release of :pypi:`humanfriendly` however in the mean time that package
  dropped support for Python 3.4.

- This explains how a simple bug fix release concerning two lines in the code
  base triggered a major version bump because compatibility is changed.

- While I was at it I set up Python 3.8 testing on Travis CI which seems to
  work fine, so I've documented Python 3.8 as compatible. Python 3.9 seems
  to be a whole other story, I'll get to that soon.

.. _Release 8.0: https://github.com/xolox/python-rotate-backups/compare/7.2...8.0
.. _#24: https://github.com/xolox/python-rotate-backups/issues/24

`Release 7.2`_ (2020-02-14)
---------------------------

Merged pull request `#23`_ which makes it possible to customize the regular
expression that's used to match timestamps in filenames using a new command
line option ``rotate-backups --timestamp-pattern``.

The pull request wasn't exactly complete (the code couldn't have run as
written, although it showed the general idea clear enough) so I decided to
treat `#23`_ as more of a feature suggestion. However there was no reason no to
merge the pull request and use it as a base for my changes, hence why I decided
to do so despite rewriting the code.

Changes from the pull request:

- Renamed ``timestamp`` to :attr:`~rotate_backups.RotateBackups.timestamp_pattern`
  to make it less ambiguous.

- Added validation that custom patterns provided by callers define
  named capture groups corresponding to the required date components
  (year, month and day).

- Rewrote the mapping from capture groups to :class:`datetime.datetime`
  arguments as follows:

  - Previously positional :class:`datetime.datetime` arguments were used
    which depended on the order of capture groups in the hard coded
    regular expression pattern to function correctly.

  - Now that users can define their own patterns, this is no longer a
    reasonable approach. As such the code now constructs and passes a
    dictionary of keyword arguments to :class:`datetime.datetime`.

- Updated the documentation and the command line interface usage message to
  describe the new command line option and configuration file option.

- Added tests for the new behavior.

.. _Release 7.2: https://github.com/xolox/python-rotate-backups/compare/7.1...7.2
.. _#23: https://github.com/xolox/python-rotate-backups/pull/23

`Release 7.1`_ (2020-02-13)
---------------------------

- Make it possibly to disable system logging using ``rotate-backups
  --syslog=false`` (fixes `#20`_).

- Explicitly support numeric :man:`ionice` classes (as required by
  :man:`busybox` and suggested in `#14`_):

  - This follows up on a pull request to :pypi:`executor` (a dependency of
    :pypi:`rotate-backups`) that was merged in 2018.

  - Since that pull request was merged this new "feature" has been implicitly
    supported by :pypi:`rotate-backups` by upgrading the installed version of
    the :pypi:`executor` package, however this probably wasn't clear to anyone
    who's not a Python developer 😇.

  - I've now merged pull request `#14`_ which adds a test to confirm that
    numeric :man:`ionice` classes are supported.

  - I also bumped the :pypi:`executor` requirement and updated the usage
    instructions to point out that numeric :man:`ionice` classes are now
    supported.

.. _Release 7.1: https://github.com/xolox/python-rotate-backups/compare/7.0...7.1
.. _#20: https://github.com/xolox/python-rotate-backups/issues/20
.. _#14: https://github.com/xolox/python-rotate-backups/pull/14

`Release 7.0`_ (2020-02-12)
---------------------------

**Significant changes:**

- Sanity checks are done to ensure the directory with backups exists, is
  readable and is writable. However `#18`_ made it clear that such sanity
  checks can misjudge the situation, which made me realize an escape hatch
  should be provided. The new ``--force`` option makes ``rotate-backups``
  continue even if sanity checks fail.

- Skip the sanity check that the directory with backups is writable when the
  ``--removal-command`` option is given (because custom removal commands imply
  custom semantics, see `#18`_ for an example).

**Miscellaneous changes:**

- Start testing on Python 3.7 and document compatibility.
- Dropped Python 2.6 (I don't think anyone still cares about this 😉).
- Copied Travis CI workarounds for MacOS from :pypi:`humanfriendly`.
- Updated ``Makefile`` to use Python 3 for local development.
- Bumped copyright to 2020.

.. _Release 7.0: https://github.com/xolox/python-rotate-backups/compare/6.0...7.0
.. _#18: https://github.com/xolox/python-rotate-backups/issues/18

`Release 6.0`_ (2018-08-03)
---------------------------

This is a bug fix release that changes the behavior of the program, and because
`rotate-backups` involves the deletion of important files I'm considering this

rotate_backups/cli.py
# rotate-backups: Simple command line interface for backup rotation.
#
# Author: Peter Odding <peter@peterodding.com>
# Last Change: May 17, 2020
# URL: https://github.com/xolox/python-rotate-backups

"""
Usage: rotate-backups [OPTIONS] [DIRECTORY, ..]

Easy rotation of backups based on the Python package by the same name.

To use this program you specify a rotation scheme via (a combination of) the
--hourly, --daily, --weekly, --monthly and/or --yearly options and the
directory (or directories) containing backups to rotate as one or more
positional arguments.

You can rotate backups on a remote system over SSH by prefixing a DIRECTORY
with an SSH alias and separating the two with a colon (similar to how rsync
accepts remote locations).

Instead of specifying directories and a rotation scheme on the command line you
can also add them to a configuration file. For more details refer to the online
documentation (see also the --config option).

Please use the --dry-run option to test the effect of the specified rotation
scheme before letting this program loose on your precious backups! If you don't
test the results using the dry run mode and this program eats more backups than
intended you have no right to complain ;-).

Supported options:

  -M, --minutely=COUNT

    In a literal sense this option sets the number of "backups per minute" to
    preserve during rotation. For most use cases that doesn't make a lot of
    sense :-) but you can combine the --minutely and --relaxed options to
    preserve more than one backup per hour.  Refer to the usage of the -H,
    --hourly option for details about COUNT.

  -H, --hourly=COUNT

    Set the number of hourly backups to preserve during rotation:

    - If COUNT is a number it gives the number of hourly backups to preserve,
      starting from the most recent hourly backup and counting back in time.
    - Alternatively you can provide an expression that will be evaluated to get
      a number (e.g. if COUNT is `7 * 2' the result would be 14).
    - You can also pass `always' for COUNT, in this case all hourly backups are
      preserved.
    - By default no hourly backups are preserved.

  -d, --daily=COUNT

    Set the number of daily backups to preserve during rotation. Refer to the
    usage of the -H, --hourly option for details about COUNT.

  -w, --weekly=COUNT

    Set the number of weekly backups to preserve during rotation. Refer to the
    usage of the -H, --hourly option for details about COUNT.

  -m, --monthly=COUNT

    Set the number of monthly backups to preserve during rotation. Refer to the
    usage of the -H, --hourly option for details about COUNT.

  -y, --yearly=COUNT

    Set the number of yearly backups to preserve during rotation. Refer to the
    usage of the -H, --hourly option for details about COUNT.

  -t, --timestamp-pattern=PATTERN

    Customize the regular expression pattern that is used to match and extract
    timestamps from filenames. PATTERN is expected to be a Python compatible
    regular expression that must define the named capture groups 'year',
    'month' and 'day' and may define 'hour', 'minute' and 'second'.

  -I, --include=PATTERN

    Only process backups that match the shell pattern given by PATTERN. This
    argument can be repeated. Make sure to quote PATTERN so the shell doesn't
    expand the pattern before it's received by rotate-backups.

  -x, --exclude=PATTERN

    Don't process backups that match the shell pattern given by PATTERN. This
    argument can be repeated. Make sure to quote PATTERN so the shell doesn't
    expand the pattern before it's received by rotate-backups.

  -j, --parallel

    Remove backups in parallel, one backup per mount point at a time. The idea
    behind this approach is that parallel rotation is most useful when the
    files to be removed are on different disks and so multiple devices can be
    utilized at the same time.

    Because mount points are per system the -j, --parallel option will also
    parallelize over backups located on multiple remote systems.

  -p, --prefer-recent

    By default the first (oldest) backup in each time slot is preserved. If
    you'd prefer to keep the most recent backup in each time slot instead then
    this option is for you.

  -r, --relaxed

    By default the time window for each rotation scheme is enforced (this is
    referred to as strict rotation) but the -r, --relaxed option can be used
    to alter this behavior. The easiest way to explain the difference between
    strict and relaxed rotation is using an example:

    - When using strict rotation and the number of hourly backups to preserve
      is three, only backups created in the relevant time window (the hour of
      the most recent backup and the two hours leading up to that) will match
      the hourly frequency.

    - When using relaxed rotation the three most recent backups will all match
      the hourly frequency (and thus be preserved), regardless of the
      calculated time window.

    If the explanation above is not clear enough, here's a simple way to decide
    whether you want to customize this behavior or not:

    - If your backups are created at regular intervals and you never miss an
      interval then strict rotation (the default) is probably the best choice.

    - If your backups are created at irregular intervals then you may want to
      use the -r, --relaxed option in order to preserve more backups.

  -i, --ionice=CLASS

    Use the `ionice' program to set the I/O scheduling class and priority of
    the `rm' invocations used to remove backups. CLASS is expected to be one of
    the values `idle' (3), `best-effort' (2) or `realtime' (1). Refer to the
    man page of the `ionice' program for details about these values. The
    numeric values are required by the 'busybox' implementation of 'ionice'.

  -c, --config=FILENAME

    Load configuration from FILENAME. If this option isn't given the following
    default locations are searched for configuration files:

    - /etc/rotate-backups.ini and /etc/rotate-backups.d/*.ini
    - ~/.rotate-backups.ini and ~/.rotate-backups.d/*.ini
    - ~/.config/rotate-backups.ini and ~/.config/rotate-backups.d/*.ini

    Any available configuration files are loaded in the order given above, so
    that sections in user-specific configuration files override sections by the
    same name in system-wide configuration files. For more details refer to the
    online documentation.

  -C, --removal-command=CMD

    Change the command used to remove backups. The value of CMD defaults to
    ``rm -fR``. This choice was made because it works regardless of whether
    "backups to be rotated" are files or directories or a mixture of both.

    As an example of why you might want to change this, CephFS snapshots are
    represented as regular directory trees that can be deleted at once with a
    single 'rmdir' command (even though according to POSIX semantics this
    command should refuse to remove nonempty directories, but I digress).

  -u, --use-sudo

    Enable the use of `sudo' to rotate backups in directories that are not
    readable and/or writable for the current user (or the user logged in to a
    remote system over SSH).

  -S, --syslog=CHOICE

    Explicitly enable or disable system logging instead of letting the program
    figure out what to do. The values '1', 'yes', 'true' and 'on' enable system
    logging whereas the values '0', 'no', 'false' and 'off' disable it.

  -f, --force

    If a sanity check fails an error is reported and the program aborts. You
    can use --force to continue with backup rotation instead. Sanity checks
    are done to ensure that the given DIRECTORY exists, is readable and is
    writable. If the --removal-command option is given then the last sanity
    check (that the given location is writable) is skipped (because custom
    removal commands imply custom semantics).

  -n, --dry-run

    Don't make any changes, just print what would be done. This makes it easy
    to evaluate the impact of a rotation scheme without losing any backups.

  -v, --verbose

    Increase logging verbosity (can be repeated).

  -q, --quiet

    Decrease logging verbosity (can be repeated).

  -h, --help

    Show this message and exit.
"""

# Standard library modules.
import getopt
import shlex
import sys

# External dependencies.
import coloredlogs
from coloredlogs.syslog import enable_system_logging
from executor import validate_ionice_class
from humanfriendly import coerce_boolean, parse_path
from humanfriendly.compat import on_windows
from humanfriendly.terminal import usage
from humanfriendly.text import pluralize
from verboselogs import VerboseLogger

# Modules included in our package.
from rotate_backups import (
    RotateBackups,
    coerce_location,
    coerce_retention_period,
    load_config_file,
)

# Initialize a logger.
logger = VerboseLogger(__name__)


def main():
    """Command line interface for the ``rotate-backups`` program."""
    coloredlogs.install()
    # Command line option defaults.
    rotation_scheme = {}
    kw = dict(include_list=[], exclude_list=[])
    parallel = False
    use_sudo = False
    use_syslog = (not on_windows())
    # Internal state.
    selected_locations = []
    # Parse the command line arguments.
    try:
        options, arguments = getopt.getopt(sys.argv[1:], 'M:H:d:w:m:y:t:I:x:jpri:c:C:uS:fnvqh', [
            'minutely=', 'hourly=', 'daily=', 'weekly=', 'monthly=', 'yearly=',
            'timestamp-pattern=', 'include=', 'exclude=', 'parallel',
            'prefer-recent', 'relaxed', 'ionice=', 'config=',
            'removal-command=', 'use-sudo', 'syslog=', 'force',
            'dry-run', 'verbose', 'quiet', 'help',
        ])
        for option, value in options:
            if option in ('-M', '--minutely'):
                rotation_scheme['minutely'] = coerce_retention_period(value)
            elif option in ('-H', '--hourly'):
                rotation_scheme['hourly'] = coerce_retention_period(value)
            elif option in ('-d', '--daily'):
                rotation_scheme['daily'] = coerce_retention_period(value)
            elif option in ('-w', '--weekly'):
                rotation_scheme['weekly'] = coerce_retention_period(value)
            elif option in ('-m', '--monthly'):
                rotation_scheme['monthly'] = coerce_retention_period(value)
            elif option in ('-y', '--yearly'):
                rotation_scheme['yearly'] = coerce_retention_period(value)
            elif option in ('-t', '--timestamp-pattern'):
                kw['timestamp_pattern'] = value
            elif option in ('-I', '--include'):
                kw['include_list'].append(value)
            elif option in ('-x', '--exclude'):
                kw['exclude_list'].append(value)
            elif option in ('-j', '--parallel'):
                parallel = True
            elif option in ('-p', '--prefer-recent'):
                kw['prefer_recent'] = True
            elif option in ('-r', '--relaxed'):
                kw['strict'] = False
            elif option in ('-i', '--ionice'):
                value = validate_ionice_class(value.lower().strip())
                kw['io_scheduling_class'] = value
            elif option in ('-c', '--config'):
                kw['config_file'] = parse_path(value)
            elif option in ('-C', '--removal-command'):
                removal_command = shlex.split(value)
                logger.info("Using custom removal command: %s", removal_command)
                kw['removal_command'] = removal_command
            elif option in ('-u', '--use-sudo'):
                use_sudo = True
            elif option in ('-S', '--syslog'):
                use_syslog = coerce_boolean(value)
            elif option in ('-f', '--force'):
                kw['force'] = True
            elif option in ('-n', '--dry-run'):
                logger.info("Performing a dry run (because of %s option) ..", option)
                kw['dry_run'] = True
            elif option in ('-v', '--verbose'):
                coloredlogs.increase_verbosity()
            elif option in ('-q', '--quiet'):
                coloredlogs.decrease_verbosity()
            elif option in ('-h', '--help'):
                usage(__doc__)
                return
            else:
                assert False, "Unhandled option! (programming error)"
        if use_syslog:
            enable_system_logging()
        if rotation_scheme:
            logger.verbose("Rotation scheme defined on command line: %s", rotation_scheme)
        if arguments:
            # Rotation of the locations given on the command line.
            location_source = 'command line arguments'
            selected_locations.extend(coerce_location(value, sudo=use_sudo) for value in arguments)
        else:
            # Rotation of all configured locations.
            location_source = 'configuration file'
            selected_locations.extend(
                location for location, rotation_scheme, options
                in load_config_file(configuration_file=kw.get('config_file'), expand=True)
            )
        # Inform the user which location(s) will be rotated.
        if selected_locations:
            logger.verbose("Selected %s based on %s:",
                           pluralize(len(selected_locations), "location"),
                           location_source)
            for number, location in enumerate(selected_locations, start=1):
                logger.verbose(" %i. %s", number, location)
        else:
            # Show the usage message when no directories are given nor configured.
            logger.verbose("No location(s) to rotate selected.")
            usage(__doc__)
            return
    except Exception as e:
        logger.error("%s", e)
        sys.exit(1)
    # Rotate the backups in the selected directories.
    program = RotateBackups(rotation_scheme, **kw)
    if parallel:
        program.rotate_concurrent(*selected_locations)
    else:
        for location in selected_locations:
            program.rotate_backups(location)
